File: 03/python/main.py
from collections import Counter


def to_decimal(bitstring):
    with_place = enumerate(reversed(bitstring))
    return sum(bit * (2 ** place) for place, bit in with_place)


def part1(bitstrings):
    def most_common_bit(bits):
        counts = Counter(bits)
        return 1 if counts[1] >= counts[0] else 0

    def complement(bitstring):
        return [1 - bit for bit in bitstring]

    gamma = [most_common_bit(bits) for bits in zip(*bitstrings)]
    epsilon = complement(gamma)
    return to_decimal(gamma) * to_decimal(epsilon)


def part2(bitstrings):
    def split_strings_by_ith_bit(bitstrings, i):
        zeros, ones = [], []
        for bits in bitstrings:
            (zeros if bits[i] == 0 else ones).append(bits)
        return zeros, ones

    def oxygen_generator(candidates, i=0):
        if len(candidates) == 1:
            return candidates[0]
        zeros, ones = split_strings_by_ith_bit(candidates, i)
        return oxygen_generator(max(ones, zeros, key=len), i + 1)

    def co2_scrubber(candidates, i=0):
        if len(candidates) == 1:
            return candidates[0]
        zeros, ones = split_strings_by_ith_bit(candidates, i)
        return co2_scrubber(min(zeros, ones, key=len), i + 1)

    return to_decimal(oxygen_generator(bitstrings)) * to_decimal(
        co2_scrubber(bitstrings)
    )


with open("../input.txt") as f:
    bitstrings = [[int(b) for b in line.strip()] for line in f]
    print(part1(bitstrings))
    print(part2(bitstrings))

File: 03/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.partitioningBy;;

class Main {
    public static void main(String[] args) throws IOException {
        var bitstrings = Files.readAllLines(Path.of("../input.txt"));
        System.out.println(part1(bitstrings));
        System.out.println(part2(bitstrings));
    }

    static int part1(List<String> bitstrings) {
        var n = bitstrings.get(0).length();
        var gammaBits = new char[n];
        Arrays.fill(gammaBits, '0');
        IntStream.range(0, n).filter(i -> numOnesSet(bitstrings, i) > bitstrings.size() / 2)
                .forEach(i -> gammaBits[i] = '1');
        var gamma = new String(gammaBits);
        var epsilon = complement(gamma);
        return toInt(gamma) * toInt(epsilon);
    }

    static int part2(List<String> bitstrings) {
        var o2 = filterBySelector(bitstrings, 0, (zeros, ones) -> ones.size() >= zeros.size() ? ones : zeros);
        var co2 = filterBySelector(bitstrings, 0, (zeros, ones) -> zeros.size() <= ones.size() ? zeros : ones);
        return toInt(o2) * toInt(co2);
    }

    static String filterBySelector(List<String> candidates, int i,
            BiFunction<List<String>, List<String>, List<String>> selector) {
        if (candidates.size() == 1) {
            return candidates.get(0);
        }
        Map<Boolean, List<String>> partitions = candidates.stream()
                .collect(partitioningBy(bits -> bits.charAt(i) == '1'));
        return filterBySelector(selector.apply(partitions.get(false), partitions.get(true)), i + 1, selector);
    }

    static String complement(String bits) {
        var result = new StringBuilder(bits.length());
        for (char c : bits.toCharArray()) {
            result.append(c == '1' ? '0' : '1');
        }
        return result.toString();
    }

    static long numOnesSet(List<String> bitstrings, int idx) {
        return bitstrings.stream().filter(bits -> bits.charAt(idx) == '1').count();
    }

    static int toInt(String bitstring) {
        var n = bitstring.length();
        return IntStream.range(0, n).filter(i -> bitstring.charAt(i) == '1').map(i -> 1 << (n - i - 1)).sum();
    }
}